beacon-chain/execution/testing/genesis.go

```diff
@@ -0,0 +1,202 @@
+// like in an e2e test. The parameters are minimal but the full value is returned unmarshaled so that it can be
+// customized as desired.
+func GethTestnetGenesis(genesisTime uint64, cfg *clparams.BeaconChainConfig) *core.Genesis {
+	cc := &params.ChainConfig{
+		ChainID:                       big.NewInt(DefaultTestChainId),
+		HomesteadBlock:                bigz,
+		DAOForkBlock:                  bigz,
+		EIP150Block:                   bigz,
+		EIP155Block:                   bigz,
+		EIP158Block:                   bigz,
+		ByzantiumBlock:                bigz,
+		ConstantinopleBlock:           bigz,
+		PetersburgBlock:               bigz,
+		IstanbulBlock:                 bigz,
+		MuirGlacierBlock:              bigz,
+		BerlinBlock:                   bigz,
+		LondonBlock:                   bigz,
+		ArrowGlacierBlock:             bigz,
+		GrayGlacierBlock:              bigz,
+		MergeNetsplitBlock:            bigz,
+		TerminalTotalDifficulty:       bigz,
+		TerminalTotalDifficultyPassed: false,
+	}
+	da := DefaultDepositContractAllocation(cfg.DepositContractAddress)
+	ma := MinerAllocation()
+	return &core.Genesis{
+		Config:     cc,
+		Nonce:      0, // overridden for authorized signer votes in clique, so we should leave it empty?
+		Timestamp:  genesisTime,
+		ExtraData:  []byte(DefaultCliqueSigner),
+		GasLimit:   math.MaxUint64 >> 1, // shift 1 back from the max, just in case
+		Difficulty: common.HexToHash(DefaultDifficulty).Big(),
+		Mixhash:    common.HexToHash(DefaultMixhash),
+		Coinbase:   common.HexToAddress(DefaultCoinbase),
+		Alloc: core.GenesisAlloc{
+			da.Address: da.Account,
+			ma.Address: ma.Account,
+		},
+		ParentHash: common.HexToHash(DefaultParenthash),
+	}
+}
+
+type DepositAllocation struct {
+	Address common.Address
+	Account core.GenesisAccount
+}
+
+func MinerAllocation() DepositAllocation {
+	return DepositAllocation{
+		Address: common.HexToAddress(DefaultMinerAddress),
+		Account: core.GenesisAccount{
+			Balance: minerBalance,
+		},
+	}
+}
+
+func DefaultDepositContractAllocation(contractAddress string) DepositAllocation {
+	s := make(map[common.Hash]common.Hash)
+	for k, v := range DefaultDepositContractStorage {
+		s[common.HexToHash(k)] = common.HexToHash(v)
+	}
+	codeBytes, err := hexutil.Decode(DepositContractCode)
+	if err != nil {
+		panic(err)
+	}
+	return DepositAllocation{
+		Address: common.HexToAddress(contractAddress),
+		Account: core.GenesisAccount{
+			Code:    codeBytes,
+			Storage: s,
+			Balance: bigz,
+			Nonce:   deterministicNonce(0),
+		},
+	}
+}
+
+func deterministicNonce(i uint64) uint64 {
+	return math.MaxUint64/2 + i
+}
+
+var ErrMalformedGenesisJson = errors.New("encoded genesis.json doesn't have expected layout")
+
+// TerribleMarshalHack calls json.Marshal and also replaces the contract address string with a '0x' prefixed version,
+// because for some reason geth broke their json marshaling.
+func TerribleMarshalHack(g *core.Genesis, addresses ...string) ([]byte, error) {
```

Comment:
for some reason json.Marshal on the genesis.json struct does not marshal the `common.Address` values as `0x` prefixed hex strings as is expected by geth init. So this dirty hack takes a list of addresses in the `alloc` field of the block to fix when marshaling.

Reply:
That is odd, how does geth then marshal this ? It should also be using the core genesis struct to marshal into a json config.

Reply:
Yeah I was surprised because the `common.Address` type does have a `MarshalText`[1], which uses their `hexutil.Bytes` type's `MarshalText`[2], which should clearly give us a `0x` prefixed string. Maybe core geth never needs to marshal this value, since it basically represents an on-disk configuration? I'll try to figure out what's going on. Do you consider this a blocker to merge the PR if we can't resolve it without this kind of hack?

[1] https://github.com/ethereum/go-ethereum/blob/master/common/types.go#L313
[2] https://github.com/ethereum/go-ethereum/blob/master/common/hexutil/json.go#L40

Reply:
Not really, just that if geth has a way to represent it as json then we don't need to hack it ourselves